### DISCUSSION TITLE
First Class FreeBSD support

### DISCUSSION THREAD

**Discussion by 6543** [woodpecker-ci/woodpecker#2245] [Request a Feature]
there exist https://wiki.freebsd.org/Docker

and [Containers](https://wiki.freebsd.org/Containers) exist since a long time, called "jails".

so it would be nice to have first class support for FreeBSD with our core infrastructure: server, agent, cli & plugin-git

This would set us also apart from other CI/CDs and put old ones like Jenkins out of work ;)

**Comment by namgo** [comment]
I'm a FreeBSD user, I was not aware of attempts made to get docker working so that's exciting! My impression from the Docker article is that it's not as stable as it might appear (2020 is fairly recent in FreeBSD development, and I don't know how many people are using it), although I might be wrong.

FreeBSD Linux binary compat is not guaranteed to work in all cases (https://docs.freebsd.org/en/books/handbook/linuxemu/ intro) and can take a bit of fiddling.

As mentioned in https://github.com/woodpecker-ci/woodpecker/discussions/2846 I've so far built agent and cli without issue, they appear to run fine. Server's UI took some work. I intend to run server on baremetal (or jailed) and agents on another system as isolated users.

**Reply by b8b** [reply]
Compiling the server go backend was also flawless?
I think it should also be possible to compile the UI on Linux as a workaround?

Docker has been ported to FreeBSD years ago but I think that most ideas in docker really revolve around Linux (e.g. it's lack of a stable base). It has been deleted at some point and I don't know about the current state.

In the meanwhile, podman and buildah are also available on FreeBSD now. Probably those work out of the box with woodpecker agent built for FreeBSD?

However, personally, I believe that builds should run inside VMs. I have created a provisioning tool that is able to spawn systems using bhyve or qemu and I want to integrate that with woodpecker. Hopefully, I find time for that next year.

**Reply by namgo** [reply]
> Compiling the server go backend was also flawless?

If I recall correctly, the only problem I had with the backend itself was of my own doing! I messed up my go/bin paths during the process. Otherwise, yeah it just worked.

The initial project that I had in mind for FreeBSD woodpecker-ci runs was scrapped, so I have yet to *run* the system yet, so take what I say with a grain of salt!

> I believe that builds should run inside VMs

I'm with you on this, FreeBSD as a host OS is great if your CI/CD will extend only to FreeBSD, but that's pretty rare. VMs in most systems are reasonably fast now!

**Comment by theoden8** [comment]
There's also qemu

* https://github.com/vmactions
* https://github.com/cross-platform-actions

Sometimes it takes its sweet time to start but otherwise quite reliable.

**Reply by 6543** [reply]
the way the actions do it, should be easy to ported it as woodpecker plugin ...

**Comment by yonas** [comment]
> This would set us also apart from other CI/CDs and put old ones like Jenkins out of work ;)

Not to distract from the main point of this thread, but yeah, we need to retire Jenkins. I think this project is a great contender for the most popular CI in the open source community.
